### PR TITLE
Wb 131 enable copy uv of manual u vs

### PR DESCRIPTION
**Pull request by MartinImgSpc**
Allow CopyPaste of manual uv. Ensure that after a paste the destination is set to manual/auto based on the source. Added the ability to CopyPaste directly in the UV editor.

### CODE REVIEW

File: Editor/EditorCore/UVEditor.cs [hunk 1]
```diff
@@ -718,44 +718,81 @@ void OnScreenResize()
...
 
-                ProBuilderEditor.instance.GetFirstSelectedFace(out firstObj, out source);
+                bool destinationWasManualUV = targetFace.manualUV;
+                bool sourceWasManualUV = source[0].manualUV;
+                if (sourceWasManualUV == true)
+                {
+                    //We need to convert it to auto
+                    UVEditing.SetAutoUV(firstObj, source, true);
+                    pb.Refresh();
+                    pb.Optimize();
```
Comment (karl-) [inline Editor/EditorCore/UVEditor.cs]: There's no need to call Refresh and Optimize here. They only need to be called when we're done editing the mesh.

File: Editor/EditorCore/UVEditor.cs [hunk 2]
```diff
@@ -718,44 +718,81 @@ void OnScreenResize()
...
+        bool CopyFaceUVSettings(ProBuilderMesh pb, Face targetFace)
         {
-            Event e = Event.current;
+            // get first selected UV face
+            ProBuilderMesh firstObj;
+            Face[] source = new Face[1];
 
-            // Copy UV settings
-            if (e.modifiers == (EventModifiers.Control | EventModifiers.Shift))
+            ProBuilderEditor.instance.GetFirstSelectedFace(out firstObj, out source[0]);
```
Comment (karl-) [inline Editor/EditorCore/UVEditor.cs]: I think this should instead check `MeshSelection.activeMesh` and `MeshSelection.activeFace`.

File: Editor/EditorCore/UVEditor.cs [hunk 3]
```diff
@@ -718,44 +718,81 @@ void OnScreenResize()
...
+                    //We need to convert it back to false
+                    UVEditing.SetAutoUV(firstObj, source, false); 
+                }
 
-                    selectedFace.uv = new AutoUnwrapSettings(source.uv);
-                    selectedFace.submeshIndex = source.submeshIndex;
-                    EditorUtility.ShowNotification("Copy UV Settings");
+                pb.ToMesh();
+                pb.Refresh();
+                pb.Optimize();
```
Comment (karl-) [inline Editor/EditorCore/UVEditor.cs]: Same note as above, we only need these calls at the end of mesh edits.

File: Editor/EditorCore/UVEditor.cs [hunk 4]
```diff
@@ -1150,6 +1202,21 @@ private void SetTool_Internal(Tool tool)
...
+            {
+                if (Math.PointInPolygon(pb.texturesInternal, mpos, pb.facesInternal[i].edgesInternal.AllTriangles()))
+                {
+                    faceSelected = pb.facesInternal[i];
+                    return true;
+                }
+            }
+            faceSelected = new Face();
+            return false;
+        }
```
Comment (karl-) [inline Editor/EditorCore/UVEditor.cs]: 👍 to moving this to a method.

File: Editor/EditorCore/UVEditor.cs (hunk 1 above)
Comment (MartinImgSpc) [reply Editor/EditorCore/UVEditor.cs]: If I do not call those I get error
image

File: Editor/EditorCore/UVEditor.cs (hunk 2 above)
Comment (MartinImgSpc) [reply Editor/EditorCore/UVEditor.cs]: Changed

File: Editor/EditorCore/UVEditor.cs (hunk 3 above)
Comment (MartinImgSpc) [reply Editor/EditorCore/UVEditor.cs]: Same comment as above

File: Editor/EditorCore/UVEditor.cs (hunk 1 above)
Comment (karl-) [reply Editor/EditorCore/UVEditor.cs]: All of the ToMesh and Refresh calls can be deferred to the end of the function. I quickly tried it out here, which works: https://github.com/Unity-Technologies/com.unity.probuilder/tree/WB-131_Enable_CopyUV_OfManual_UVs_Edit

Comment (MartinImgSpc) [reply Editor/EditorCore/UVEditor.cs]: As mentioned it did not work when the source and target manual/auto did not match. Also when source was manual.

**Comment by gabrielw-us** [comment]
Problem:
1) Set Face A to Manual
2) Set Face B to Auto
3) Attempt Copy-Paste from B onto A
4) Nothing appears to happen
5) Attempt again - this time it appears to work

After much digging around, I realized that in step 3 above, the Manual face is converted to Auto, but the actual UV settings are not applied. Step 5 does that. Those two should happen in a single step (when pasted).

Copy from an Auto face to a Manual face seems to work correctly, just one step.

**Comment by MartinImgSpc** [comment]
> Problem:
> 
> 1. Set Face A to Manual
> 2. Set Face B to Auto
> 3. Attempt Copy-Paste from B onto A
> 4. Nothing appears to happen
> 5. Attempt again - this time it appears to work
> 
> After much digging around, I realized that in step 3 above, the Manual face is converted to Auto, but the actual UV settings are not applied. Step 5 does that. Those two should happen in a single step (when pasted).
> 
> Copy from an Auto face to a Manual face seems to work correctly, just one step.

@gabrielw-us good catch this should be fixed now! Thank you!

**Comment by MartinImgSpc** [comment]
@gabrielw-us  @karl- I have removed the conversion from the source face UV settings so that there is no loss during conversion. I think we should do another test pass to ensure there is no regression.

File: Runtime/MeshOperations/UV/AutoUvConversion.cs [hunk 5]
```diff
@@ -83,14 +83,34 @@ public override string ToString()
...
+        /// </summary>
+        /// <returns></returns>
+        internal static AutoUnwrapSettings GetAutoUnwrapSettings(ProBuilderMesh mesh, Face face)
+        {
+            if (!face.manualUV)
+                return new AutoUnwrapSettings(face.uv);
+
+            var trs = GetUVTransform(mesh, face);
+            var uvSettings = AutoUnwrapSettings.defaultAutoUnwrapSettings;
+            uvSettings.offset = trs.translation;
```
Comment (karl-) [inline Runtime/MeshOperations/UV/AutoUvConversion.cs]: doesn't the offset need to be negated when assigned to `AutoUnwrapSettings`?

File: Editor/EditorCore/UVEditor.cs [hunk 6]
```diff
@@ -718,44 +718,91 @@ void OnScreenResize()
...
+            bool sourceWasManualUV = source[0].manualUV;
+                
+            if (sourceWasManualUV == true)
             {
-                // get first selected Auto UV face
-                ProBuilderMesh firstObj;
-                Face source;
+                //We need to convert it to auto
+                firstObj.ToMesh();
+                UVEditing.SetAutoUV(firstObj, source, true);
```
[the remark this answers is not in the document]
Comment (MartinImgSpc) [reply Editor/EditorCore/UVEditor.cs]: You are looking at the outdated version

File: Runtime/MeshOperations/UV/AutoUvConversion.cs (hunk 5 above)
Comment (MartinImgSpc) [reply Runtime/MeshOperations/UV/AutoUvConversion.cs]: testing says otherwise

**Pull request merged**